### PR TITLE
SE: ComputeBoolConstraint for numeric comparisons

### LINKED ISSUE #2528: handling of null conditional and null coalescing operators can lead to a false positive for S2259
When an IF statement uses the null conditional rule and (or?) the null coalescing operator then the else path is not marked as cannot be null, resulting in false S2259 messages.

Eg.
if ((x?.value??0)==0)
{ // this happens if x is null, x.value is null or x.value == 0
}
else
{ // this part can't be called if x is null or x.value is null
  text = $"{x.value} items";
}

In this example, the line 
  text = $"{x.value} items";
gets a message S2259 'x' is null on at least one execution path.
Unless I'm misunderstanding something, that is not true.

### LINKED ISSUE #7142: SE: ComputeBoolConstraint for numeric comparisons
Compute `BoolConstraint` for known result of ==, !=, >, >=, <, <= conditional operators.

Building NumberConstraint from the comparison is another issue: #7143

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**
Fixes #7142
Fixes #2528

I'm not going to fix the coverage, because we need #7143 first

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.cs [hunk 1]
```diff
@@ -126,6 +131,22 @@ kind switch
             _ => null
         };
 
+    private static SymbolicConstraint BinaryNumberConstraint(BinaryOperatorKind kind, NumberConstraint left, NumberConstraint right) =>
+        kind switch
+        {
+            BinaryOperatorKind.Equals when left.IsSingleValue && right.IsSingleValue => BoolConstraint.From(left.Equals(right)),
+            BinaryOperatorKind.NotEquals when left.IsSingleValue && right.IsSingleValue => BoolConstraint.From(!left.Equals(right)),
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.cs:138]: Here, we will also need this, but there's no way to cover it yet. A note is added to 
```
BinaryOperatorKind.NotEquals when left.Min > right.Max || left.Max < right.Min => BoolConstraint.False,
```

Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.cs:137]: I prefer to add commit that will also add test, to make sure it's correct. (honestly, I made a mistake when writing it on 1st try and edited the message :D )

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.cs [hunk 2]
```diff
@@ -101,6 +101,10 @@ private static SymbolicConstraint BinaryConstraint(BinaryOperatorKind kind, Symb
         {
             return BinaryBoolConstraint(kind, leftBool == BoolConstraint.True, rightBool == BoolConstraint.True);
         }
+        else if (left?.HasConstraint<NumberConstraint>() is true && right?.HasConstraint<NumberConstraint>() is true)
+        {
+            return BinaryNumberConstraint(kind, left.Constraint<NumberConstraint>(), right.Constraint<NumberConstraint>());
+        }
```
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.cs:107]: Nice

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Binary.cs [hunk 3]
```diff
@@ -657,4 +657,64 @@ public void Binary_StringConcatenation_Compund_CS(string expression)
...
+    [DataRow("arg == 42")]
+    [DataRow("arg != 42")]
+    [DataRow("arg > 42")]
+    [DataRow("arg >= 42")]
+    [DataRow("arg < 42")]
+    [DataRow("arg <= 42")]
+    public void Binary_NumberLiteral_Unknown(string expression)
+    {
+        var code = $"""
+            var result = {expression};
```
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Binary.cs:714]: I prefer the inconsistent way. It makes it clear from far away that these cases are for unknown value of arg.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/LocksReleasedAllPaths.Monitor.Loops.cs [hunk 4]
```diff
@@ -152,5 +169,61 @@ public void Method11(bool condition, List<byte> array)
...
+            {
+                Console.WriteLine();
+            }
+
+            Monitor.Enter(obj); // Noncompliant
+            if (condition)
+            {
+                Monitor.Exit(obj);
+            }
+        }
```
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/LocksReleasedAllPaths.Monitor.Loops.cs:227]: They somewhat do. There should be at least one "integration test" for each feature, to ensure it behaves nicely and as expected with the actual rule.

We've built the engine around S2222 and you can see from the file names that `LocksReleasedAllPaths.Monitor.*.cs` are meant to test different engine aspects against a rule behavior.

We don't have to test each feature against each rule, but there should be one test on at least one place for each feature-group.

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Binary.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Tim-Pohlmann) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Binary.cs:714]: That makes sense 👍

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Constraints/NumberConstraintTest.cs [hunk 5]
```diff
@@ -82,6 +82,16 @@ public void From_Object_OtherTypes()
...
         }
 
+        [TestMethod]
+        public void IsSingleValue()
+        {
+            NumberConstraint.From(42).IsSingleValue.Should().BeTrue();
+            NumberConstraint.From(1, 1).IsSingleValue.Should().BeTrue();
+            NumberConstraint.From(null, null).IsSingleValue.Should().BeFalse();
+            NumberConstraint.From(null, 42).IsSingleValue.Should().BeFalse();
+            NumberConstraint.From(42, null).IsSingleValue.Should().BeFalse();
```
Comment (Tim-Pohlmann) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Constraints/NumberConstraintTest.cs:92]: I know coverage is not a requirement for this PR, but this makes sense to test.
```suggestion
            NumberConstraint.From(42, null).IsSingleValue.Should().BeFalse();
            NumberConstraint.Empty.IsSingleValue.Should().BeFalse();
```

**Review by Tim-Pohlmann — APPROVED** [review]
LGTM. Added a single test case.

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Constraints/NumberConstraintTest.cs (hunk 5 above)
Comment (Tim-Pohlmann) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Constraints/NumberConstraintTest.cs:92]: Oh! I think it's missing `From(1, 42)`.
Additional note: I just changed the behavior of `From(null, null)` in #7174. You should change the test to use `Empty` instead.

**Pull request merged**